### PR TITLE
RFC Proposal Path

### PR DESCRIPTION
**Pull request by prince-chrismc**
As discussed in our retro, we need to formally have a processes
- This is taken deeply from Rust, Python, and React Native existing templates

We'll need to prioritize writing the proposals that we made verbally.

My goal is to fold this into the new website's documentation under the "Developer Hub".
I might tweak the header of the template to render better on the website.

### CODE REVIEW

**Comment by chb0github** [comment]
What ticket does this address?

**Review by chb0github — DISMISSED** [review]
I don't have much of an opinion on how to define an RFC template

**Comment by prince-chrismc** [comment]
> What ticket does this address?

This was an action item from the standup ❤️

File: docs/rfc/0000-template.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/pull/0000)
+- Rust Issue: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/issues/0000)
```
Comment (erwin1) [inline docs/rfc/0000-template.md:4]: why does it say **Rust** issue here?

**Comment by erwin1** [comment]
I read through it, looks like a very good starting point. 
I just had this one minor question above.

**Comment by mseabornIBM** [comment]
I think the RFC approach is good. I also believe, as is outlined in the process, that getting consensus on ideas is an important phase of proposal. Does it make sense to have a step before your RFC process steps that is a small white paper about the idea before it is committed to the RFC process?

File: docs/rfc/0000-template.md (hunk 1 above)
Comment (prince-chrismc) [reply docs/rfc/0000-template.md:4]: I absolutely stole most of this from the Rust community 

great catch

**Comment by prince-chrismc** [comment]
> Does it make sense to have a step before your RFC process steps that is a small white paper about the idea before it is committed to the RFC process?

I tried to cover this in 

https://github.com/prince-chrismc/pyrsia-sandbox/tree/proposal-path/docs/rfc#before-creating-an-rfc

> The most common preparations for writing and submitting an RFC include talking the idea over on our official Slack channel, discussing the topic in a GitHub issue.

Perhaps is a little unclear or hidden?

**Comment by prince-chrismc** [comment]
> Perhaps is a little unclear or hidden?

Looking it over, i felt it could be improved!

https://github.com/prince-chrismc/pyrsia-sandbox/blob/proposal-path/docs/rfc/readme.md#summary

Thanks for the input 👏

**Review by erwin1 — APPROVED** [review]

**Review by efrisch — DISMISSED** [review]
lgtm. 👍

**Comment by betarelease** [comment]
I was reading through this and the references from RUST community. 

For our small team and purpose this is very detailed and has too many required sections.

I would like this to be simplified as 

TL;DR
What is the Issue?
Solutions available
Proposed Solution 
Comments

And provide references in case we are making such a fundamental change that other sections may make sense and let the proposer choose to add sections.


In general reduce the barrier for people to submit proposals, but give them a little bit of structure.

**Comment by prince-chrismc** [comment]
@betarelease it's unlear if you are referring to the template of the proposal workflow 😕 

Which file (or both) so I take that feedback into account for?

File: docs/rfc/readme.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+a consensus among the Pyrsia community. See [when to follow this process](#when-you-need-to-follow-this-process) below.
+
+#### Summary
+
+The process can be broken down as follows:
+
+1. [Reach out on community forms and open an issue](#before-creating-an-rfc)
+2. [Follow the template and open a PR](#the-proposal-process)
+3. [Attend our meeting and promote your proposal](#approach-the-community)
+4. [Schedual a review meeting](#reviewing-a-proposal)
```
Comment (betarelease) [inline docs/rfc/readme.md:16]: `Schedule`

File: docs/rfc/readme.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+## The Proposal Process
+
+In short, to get a major feature added, one must first get the RFC merged into the RFC repository as a markdown file. At that point the RFC is "active" and may
+be implemented.
+
+- Fork this repository [pyrsia/pyrsia](https://github.com/pyrsia/pyrsia/fork)
+- Copy `docs/rfc/0000-template.md` to `docs/rfc/0000-my-feature.md`
+  - where 'my-feature' is the title in kebab case; don't assign a number yet.
+- Fill in the proposal. Put care into the details: **Proposals that do not present convincing motivation, demonstrate understanding of the impact of the design, or are disingenuous about the drawbacks or alternatives tend to be poorly-received**.
```
Comment (betarelease) [inline docs/rfc/readme.md:69]: I agree that we should ask people to put in sincere effort.  But not everyone proposing will have deep understanding and having these constraints will be discouraging. I would rephrase as

`Please put in enough time and research in putting a proposal together and we strongly encourage getting some feedback from part of the team before making your proposal final.`

File: docs/rfc/readme.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+- Submit a pull request. As a pull request, the proposal will receive feedback from the larger community, and the author should be prepared to revise it in response.
+
+### Approach the Community
+
+Build consensus and integrate feedback. Proposals that have broad support are much more likely to make progress than those that don't receive any comments.
+
+### Reviewing a Proposal
+
+Periodically, the team will attempt to review the active proposals. We try to dicuss proposals at the bi-weekly team
+["Architecture"](https://pyrsia.io/events/) meeting, we schedual additional meetings as need. Actions are recorded in the meeting minutes.
```
Comment (betarelease) [inline docs/rfc/readme.md:79]: `schedule`

File: docs/rfc/readme.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+Every accepted feature should have a core team champion, who will represent the feature and its progress.
```
Comment (betarelease) [inline docs/rfc/readme.md:80]: Line 80 is not a required IMO.

File: docs/rfc/readme.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+### Proposal resolution
+
+Eventually, the team will decide whether the proposal is a candidate for adoption.
+
```
Comment (betarelease) [inline docs/rfc/readme.md:85]: ... and the decision will be recorded which will specify how the proposal becomes actionable.


We can remove lines 86/87/88

File: docs/rfc/readme.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+## Implementing a proposal
+
+The author of a proposal is not obligated to implement it. Of course, the proposal author (like any other community member) is welcome to post an implementation for review.
+
+***
+
+#### Inspirations
+
+This has been havily inspired by other community driven projects and their RFC process.
```
Comment (betarelease) [inline docs/rfc/readme.md:105]: I would like to rephrase this as, 

`This has been derived from other community driven projects. But in the end the changes to this process can be proposed to the team and this process is open to being updated.`

File: docs/rfc/readme.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+## Why are RFCs needed?
+
+An RFCs is a design document providing information to the Pyrsia community. The RFC should provide a concise technical specification of
+the feature and a rationale for the feature. Strong RFCs included detailed research on the idea and may be accompanied by a proof of concept
+
+We intend RFCs to be the primary mechanisms for proposing major new features, for collecting community input on an issue, and for documenting
+the design decisions that have gone into Pyrsia.
+
+The RFC author is responsible for building consensus within the community and documenting dissenting opinions.
```
Comment (betarelease) [inline docs/rfc/readme.md:28]: This is not how we would like to drive this forward. The RFC author should be encouraged to propose and once proposal is being discussed to facilitate the discussion. 

The team can figure out if it is actionable or not (with or without a proper consensus)

**Review by betarelease — COMMENTED** [review]
I added comments inline to clarify which parts of the process flow I felt were too demanding/constraining.

Similarly any language that puts too many constraints on including people into the team should be cleaned up.

File: docs/rfc/0000-template.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+Explain the proposal as if it was already included in the language and you were teaching it to another Pyrsia member. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how Pysia users should *think* about the feature, and how it should impact the way they use Pyrsia. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing Pyrsia users and new Pyrsia users.
+
+For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
```
Comment (betarelease) [inline docs/rfc/0000-template.md:24]: Not a good example since pyrsia does not have compiler internals. Should be removed or rephrased.

File: docs/rfc/0000-template.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+
+# Drawbacks
+
+Why should we *not* do this?
```
Comment (betarelease) [inline docs/rfc/0000-template.md:38]: Instead, 
what will be the loss if we do this?

Why should we not do this? does not read the same as drawbacks to me.

File: docs/rfc/0000-template.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+# Unresolved questions
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
+
+# Future possibilities
+
+Think about what the natural extension and evolution of your proposal would
```
Comment (betarelease) [inline docs/rfc/0000-template.md:66]: This paragraph is too much to ask from a proposer.

File: docs/rfc/0000-template.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+be and how it would affect the project as a whole in a holistic
+way. Try to use this section as a tool to more fully consider all possible
+interactions with the project and language in your proposal.
+Also consider how this all fits into the roadmap for the project
+and of the relevant sub-team.
+
+This is also a good place to "dump ideas", if they are out of scope for the
+RFC you are writing but otherwise related.
+
+If you have tried and cannot think of any future possibilities,
```
Comment (betarelease) [inline docs/rfc/0000-template.md:76]: Sounds non-inclusive(meaning condescending this case)

File: docs/rfc/0000-template.md [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+you may simply state that you cannot think of anything.
+
+Note that having something written down in the future-possibilities section
```
Comment (betarelease) [inline docs/rfc/0000-template.md:79]: Too much structure enforced by adding this in the footnote.

File: docs/rfc/0000-template.md [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+# Prior art
+
+Discuss prior art, both the good and the bad, in relation to this proposal.
+A few examples of what this can include are:
+
+- Does this feature exist in other package managers and what experience have their community had?
+- For community proposals: Is this done by some other community and what were their experiences with it?
+- For other teams: What lessons can we learn from what other communities have done here?
+- Papers: Are there any published papers or great posts that discuss this? If you have some relevant papers to refer to, this can serve as a more detailed theoretical background.
```
Comment (betarelease) [inline docs/rfc/0000-template.md:54]: Rephrase to `references` (In Rust language context it might be papers but for Pyrsia it is papers but more importantly real web implementation references)

**Review by betarelease — COMMENTED** [review]
I added a few more comments in places where I found the template too demanding or too strict.

File: docs/rfc/readme.md (hunk 2 above)
Comment (prince-chrismc) [reply docs/rfc/readme.md:16]:
```suggestion
4. [Schedule a review meeting](#reviewing-a-proposal)
```

File: docs/rfc/readme.md (hunk 5 above)
Comment (prince-chrismc) [reply docs/rfc/readme.md:80]:
```suggestion
```

File: docs/rfc/readme.md (hunk 7 above)
Comment (prince-chrismc) [reply docs/rfc/readme.md:105]:
```suggestion
This has been derived from other community driven projects. But in the end the changes to this process can be proposed to the team and this process is open to being updated.
```

File: docs/rfc/0000-template.md (hunk 9 above)
Comment (prince-chrismc) [reply docs/rfc/0000-template.md:24]:
```suggestion
For implementation-oriented RFCs (e.g. for peer-to-peer networking technology stack), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
```

File: docs/rfc/0000-template.md (hunk 10 above)
Comment (prince-chrismc) [reply docs/rfc/0000-template.md:38]: > Why should we not do this? does not read the same as drawbacks to me.

🤔 Thats exactly how I learnt "drawback"

> plural noun: drawbacks
> 1. a feature that renders something less acceptable; a disadvantage or problem.
>    "the main drawback of fitting catalytic converters is the cost"

*** 

@betarelease  How do you feel about 

```suggestion
Why should we *not* do this? What are the disadvantages of the propose? What negative side effects or limitation would be consequence?
```

Is that any better?

File: docs/rfc/0000-template.md (hunk 14 above)
Comment (prince-chrismc) [reply docs/rfc/0000-template.md:54]: I agree, this can be improved.

With the blockchain specifically there are tons of papers. So this made sense, but in the wider picture there needs to be more language 

👍

File: docs/rfc/0000-template.md (hunk 11 above)
Comment (prince-chrismc) [reply docs/rfc/0000-template.md:66]: I made this sectional optional but I like the idea to leave space for anything extra the author might want to add

Is that a good middle ground?

File: docs/rfc/0000-template.md (hunk 14 above)
Comment (prince-chrismc) [reply docs/rfc/0000-template.md:54]:
```suggestion
- References: Are there any published papers or great posts that discuss this? If you have some relevant research to refer to, this can serve as a more detailed theoretical background. Existing implementation references for practical background can also help fill in a proposal.
```

File: docs/rfc/readme.md (hunk 8 above)
Comment (prince-chrismc) [reply docs/rfc/readme.md:28]:
```suggestion
The RFC author is strongly encouraged to facilitate the discussion when the proposal is brought up at community events.
```

File: docs/rfc/0000-template.md [hunk 15]
```diff
@@ -0,0 +1,74 @@
+- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/pull/0000)
+- Pyrsia Issue: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/issues/0000)
```
Comment (AbhijithGanesh) [inline docs/rfc/0000-template.md:4]: is 0000 the standard way to do it?

Comment (prince-chrismc) [reply docs/rfc/0000-template.md:4]: It's just a placeholder for any number, is that unclear?

File: docs/rfc/0000-template.md (hunk 10 above)
Comment (prince-chrismc) [reply docs/rfc/0000-template.md:38]: Yes!

Comment (prince-chrismc) [reply docs/rfc/0000-template.md:38]:
```suggestion
Why should we *not* do this? What are the trade-offs? What negative side effects or limitation could be a consequence?
```

**Review by efrisch — APPROVED** [review]
👍

**Review by fishseabowl — APPROVED** [review]
LGTM 😄
